src/bindings.cpp

```diff
@@ -21,51 +22,97 @@ extern "C"
     std::unique_ptr<Player> player = nullptr;
     std::unique_ptr<Analyzer> analyzer = std::make_unique<Analyzer>(2048);
 
-    /// @brief Set a dart function to call when the sound with [handle] handle ends
-    /// @param callback this is the dart function that will be called
-    ///     when the sound ends to play.
-    ///     Must be global or a static class member:
-    ///     ```@pragma('vm:entry-point')
-    ///        void playEndedCallback(int handle) {
-    ///             // here the sound with [handle] has ended.
-    ///             // you can play again
-    ///             soLoudController.soLoudFFI.play(handle);
-    ///             // or dispose it
-    ///             soLoudController.soLoudFFI.stop(handle);
-    ///        }
-    ///     ```
-    /// @param handle the handle to the sound
-    /// @param callback this is the dart function that will be called
-    ///         when the sound ends to play
-    /// @return true if success;
-    // FFI_PLUGIN_EXPORT bool setPlayEndedCallback(void (*callback)(unsigned int), unsigned int handle)
-    // {
-    //     if (!player.get()->isInited()) return false;
-    //     ActiveSound* sound = player.get()->findByHandle(handle);
-    //     if (sound != nullptr) {
-    //         sound->playEndedCallback = callback;
-    //         return true;
-    //     }
-    //     return false;
-    // }
+    typedef void (*dartVoiceEndedCallback_t)(unsigned int *);
+    typedef void (*dartFileLoadedCallback_t)(enum PlayerErrors *, char *completeFileName, unsigned int *);
+    typedef void (*dartStateChangedCallback_t)(enum PlayerStateEvents *);
+
+    // to be used by `NativeCallable`, these functions must return void.
+    void (*dartVoiceEndedCallback)(unsigned int *) = nullptr;
+    void (*dartFileLoadedCallback)(enum PlayerErrors *, char *completeFileName, unsigned int *) = nullptr;
+    void (*dartStateChangedCallback)(enum PlayerStateEvents *) = nullptr;
+
+    FFI_PLUGIN_EXPORT void nativeFree(void *pointer) {
+        free(pointer);
+    }
+    
+    /// The callback to monitor when a voice ends.
+    ///
+    /// It is called by void `Soloud::stopVoice_internal(unsigned int aVoice)` when a voice ends.
+    void voiceEndedCallback(unsigned int *handle)
+    {
+        if (dartVoiceEndedCallback == nullptr)
+            return;
+        player->removeHandle(*handle);
+        // [n] pointer must be deleted on Dart.
+        unsigned int *n = (unsigned int *)malloc(sizeof(unsigned int *));
+        *n = *handle;
+        dartVoiceEndedCallback(n);
```

Comment:
Ah, I think I get the `nativeFree()` function now. The memory would normally be freed here after the call to `dartVoiceEndedCallback(n)`, but the pointer is somehow being used asynchronously in Dart land so we need to call `nativeFree()` from there. Am I correct?